src/server/pfs/server/driver.go

```diff
@@ -1049,6 +1049,173 @@ func (d *driver) deleteRepo(txnCtx *txnenv.TransactionContext, repo *pfs.Repo, f
+	// TODO(msteffen): Fix d.deleteAll() so that it doesn't need to delete and
+	// recreate the PPS spec repo, then uncomment this block to prevent users from
+	// deleting it and breaking their cluster
+	// if repo.Name == ppsconsts.SpecRepo {
+	// 	return errors.Errorf("cannot delete the special PPS repo %s", ppsconsts.SpecRepo)
+	// }
+	repos := d.repos.ReadWrite(txnCtx.Stm)
+	// check if 'repo' is already gone. If so, return that error. Otherwise,
+	// proceed with auth check (avoids awkward "access denied" error when calling
+	// "deleteRepo" on a repo that's already gone)
+	var existingRepoInfo pfs.RepoInfo
+	if err := repos.Get(repo.Name, &existingRepoInfo); err != nil {
+		if !col.IsErrNotFound(err) {
+			return err
+		}
+	}
+	if err := authserver.CheckIsAuthorizedInTransaction(txnCtx, repo, auth.Scope_OWNER); err != nil {
+		return err
+	}
+	repoInfo := &pfs.RepoInfo{}
+	if err := repos.Update(repo.Name, repoInfo, func() error {
+		repoInfo.Tombstone = true
+		return nil
+	}); err != nil {
+		if !col.IsErrNotFound(err) {
+			return err
+		}
+	}
+	var branchInfos []*pfs.BranchInfo
+	for _, branch := range repoInfo.Branches {
+		bi, err := d.inspectBranch(txnCtx, branch)
+		if err != nil {
+			return err
+		}
+		branchInfos = append(branchInfos, bi)
+	}
+	sort.Slice(branchInfos, func(i, j int) bool { return len(branchInfos[i].Provenance) < len(branchInfos[j].Provenance) })
+	for i := range branchInfos {
+		// delete branches from most provenance to least, that way if one
+		// branch is provenant on another (such as with stats branches) we
+		// delete them in the right order.
+		branch := branchInfos[len(branchInfos)-1-i].Branch
+		if err := d.deleteBranch(txnCtx, branch, force); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *driver) updateCommitSubvenance(txnCtx *txnenv.TransactionContext, repo *pfs.Repo, prov *pfs.CommitProvenance) error {
+	provCI := &pfs.CommitInfo{}
+	if err := d.commits(prov.Commit.Repo.Name).ReadWrite(txnCtx.Stm).Update(prov.Commit.ID, provCI, func() error {
+		subvTo := 0 // copy subvFrom to subvTo, excepting subv ranges to delete (so that they're overwritten)
+	nextSubvRange:
+		for subvFrom, subv := range provCI.Subvenance {
+			// Compute path (of commit IDs) connecting subv.Upper to subv.Lower
+			cur := subv.Upper.ID
+			path := []string{cur}
+			for cur != subv.Lower.ID {
+				// Get CommitInfo for 'cur' from etcd
+				// and traverse parent
+				curInfo := &pfs.CommitInfo{}
+				if err := d.commits(subv.Lower.Repo.Name).ReadWrite(txnCtx.Stm).Get(cur, curInfo); err != nil {
+					return errors.Wrapf(err, "error reading commitInfo for subvenant \"%s/%s\"", subv.Lower.Repo.Name, cur)
+				}
+				if curInfo.ParentCommit == nil {
+					break
+				}
+				cur = curInfo.ParentCommit.ID
+				path = append(path, cur)
+			}
+
+			// move 'subv.Upper' through parents until it points to a non-deleted commit
+			for j := range path {
+				if subv.Upper.Repo.Name != repo.Name {
```

Comment:
Yeah, so I basically just repackaged this functionality which already existed. I think your observation is correct upon closer inspection. I think the repo check is just used for skipping the update of the commit ranges that are not associated with the repo being deleted. To be honest, I would like to avoid making changes here in the interest of mitigating risk, and because this code will get thrown away in the long term (2.0). As far as your third question is concerned, I don't think a commit can have a parent from a different repo.

Reply:
Thanks for the explanation! If this is all going away in 2.0, then I'm happy to LGTM and unblock the PR, but given what you're saying, it does seem to me like it would be lower risk to remove this code, for two reasons:
- less code overall :shrug: 
- no RPCs to etcd (to get the commitInfo, l. 1176)

If the commit IDs aren't important, I think you could replace all of this with just nine lines:
```
for subvFrom, subv := range provCI.Subvenance {
  if subv.Upper.Repo.Name == repo.Name {
    continue
  }
  provCI.Subvenance[subvTo] = provCI.Subvenance[subvFrom]
  subvTo++
}
provCI.Subvenance = provCI.Subvenance[:subvTo]
return nil
```

You've been thinking through this for longer than I have, though, to prepare the PR, so I'll defer to your expertise.

Reply:
Okay, I'm convinced that this is a change we should make, and I just pushed it. Reducing the number of commits read is going to have a nice impact on the single transaction delete repo. @jdoliner would you do a quick sanity check on this change.